### PR TITLE
Refactor odenets

### PR DESCRIPTION
**Pull request by eric-c-cyr**
@steffi7574 finally got through this. Hopefully it helps clean some things up. Thanks for you help in pointing to (now) obvious improvments!

### CODE REVIEW

**Comment by eric-c-cyr** [comment]
I plan on quashing before I merge, but I think this is good to check for now.

File: torchbraid/torchbraid_callbacks.pyx [hunk 1]
```diff
@@ -82,21 +83,35 @@ cdef int my_step(braid_App app, braid_Vector ustop, braid_Vector fstop, braid_Ve
 
       tstart = 0.0
       tstop = 5.0
+      tindex = -1
       level = -1
       braid_StepStatusGetTstartTstop(status, &tstart, &tstop)
       braid_StepStatusGetLevel(status, &level)
       braid_StepStatusGetDone(status, &done)
+      braid_StepStatusGetDone(status, &done)
+      braid_StepStatusGetTIndex(status, &tindex)
```
Comment (steffi7574) [inline torchbraid/torchbraid_callbacks.pyx:92]: When called on coarser grids, this tindex is the index of the time-point on that coarse level, hence needs to be converted to the corresponding fine-grid time-index (e.g. multiply by the "(cfactor)^level" I guess)

**Review by steffi7574 — COMMENTED** [review]
I can't review the rnn-related changes, but I did go through all the other changes and left comments. There is a bug related to setting the correct time-point index, because on coarser grids, _braid_StepStatusGetTIndex return a **level-dependent** time index, see comments in torchbraid_callbacks.pyx and follow-ups in odenet_apps.py.

File: torchbraid/layer_parallel.py [hunk 2]
```diff
@@ -118,13 +118,13 @@ def __init__(self,comm,layer_block,num_steps,Tf,max_levels=1,max_iters=10,spatia
     self.dt = Tf/global_steps
   
     self.layer_block = layer_block
-    self.layer_models = [layer_block() for i in range(num_steps)]
-    self.local_layers = nn.Sequential(*self.layer_models)
 
     self.timer_manager = ContextTimerManager()
 
-    self.fwd_app = apps.ForwardODENetApp(comm,self.layer_models,num_steps,Tf,max_levels,max_iters,self.timer_manager,
+    self.fwd_app = apps.ForwardODENetApp(comm,None,num_steps,Tf,max_levels,max_iters,self.timer_manager,
```
[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply torchbraid/layer_parallel.py:124]: Done

File: torchbraid/braid_function.py [hunk 3]
```diff
@@ -88,11 +88,11 @@ def backward(ctx, grad_output):
 
     # send gradients to the right 
     grads = ctx.bwd_app.grads
-    if my_rank<num_ranks-1:
-      comm.send(grads[-1],dest=my_rank+1,tag=22)
-    if my_rank>0:
-      neighbor_model = comm.recv(source=my_rank-1,tag=22)
-      grads.insert(0,neighbor_model)
+    #if my_rank<num_ranks-1:
```
[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply torchbraid/braid_function.py:91]: done

File: torchbraid/odenet_apps.py [hunk 4]
```diff
@@ -49,40 +49,41 @@ def __init__(self,comm,layer_models,local_num_steps,Tf,max_levels,max_iters,time
...
     my_rank       = self.getMPIComm().Get_rank()
     num_ranks     = self.getMPIComm().Get_size()
     self.my_rank = my_rank
     self.layer_block = layer_block
 
-    # add a sentinal at the end
-    self.layer_models.append(None)
+    sys.stdout.flush()
+    owned_layers = self.end_layer-self.start_layer+1
+    if my_rank==num_ranks-1:
```
[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply torchbraid/odenet_apps.py:62]: done

File: torchbraid/odenet_apps.py [hunk 5]
```diff
@@ -174,53 +155,37 @@ def eval(self,y,tstart,tstop,level,done,x=None):
...
-      q = dt*layer(t_x)
-      t_y.add_(q)
-      del q
-    # end in_place_eval
-
     # there are two paths by which eval is called:
     #  1. x is a BraidVector: my step has called this method
     #  2. x is a torch tensor: called internally (probably for the adjoint) 
 
     if isinstance(y,BraidVector):
```
[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply torchbraid/odenet_apps.py:162]: done

**Comment by eric-c-cyr** [comment]
@steffi7574 Would you suggest I simply use the global index criterion instead of the local one?  I could use `getGlobalTimeIndex` instead of `getTimeStepIndex`. I actually ended up doing that in the recurrent neural network (its strange that this caused issue there, but not in odenet)

File: torchbraid/odenet_apps.py [hunk 6]
```diff
@@ -174,53 +155,37 @@ def eval(self,y,tstart,tstop,level,done,x=None):
...
+        del q
+
+      #if y.getSendFlag():
+      #  self.clearTempLayerWeights()
+#
+#        y.releaseWeightTensors()
+#        y.setSendFlag(False)
+#      # wipe out any sent information
+
+      tstop_index = self.getTimeStepIndex()+1 # get end time stepl
```
[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply torchbraid/odenet_apps.py]: Oh, I just fully digested this. Yes. I wlil use getGlobalTimeIndex

File: torchbraid/odenet_apps.py [hunk 7]
```diff
@@ -230,20 +195,26 @@ def getPrimalWithGrad(self,tstart,tstop,level):
...
-    layer = self.getLayer(tstart,tstop,level)
 
     b_x = self.getUVector(0,tstart)
-    t_x = b_x.tensor()
 
-    self.setLayerWeights(tstart,tstop,level,b_x.weightTensors())
+    if ts_index<len(self.layer_models):
+      layer = self.layer_models[ts_index]
+    else:
+      layer = self.temp_layer
```
[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply torchbraid/odenet_apps.py:205]: Okay, working this point as well.

**Comment by eric-c-cyr** [comment]
@steffi7574 I think this is cleaned up more now. Thanks for pushing on the fixes. I believe I've resolved all your points. Let me know.

File: torchbraid/odenet_apps.py [hunk 8]
```diff
@@ -174,53 +153,37 @@ def eval(self,y,tstart,tstop,level,done,x=None):
...
-        t_x = t_y
-      else:
-        t_y.copy_(t_x)
-
-      q = dt*layer(t_x)
-      t_y.add_(q)
-      del q
-    # end in_place_eval
-
     # there are two paths by which eval is called:
```
Comment (steffi7574) [inline torchbraid/odenet_apps.py:156]: This comments is confusing now, and probably should be removed. As far as I see it, there is not anymore two path which call eval, right?

File: torchbraid/odenet_apps.py [hunk 9]
```diff
@@ -174,53 +153,37 @@ def eval(self,y,tstart,tstop,level,done,x=None):
...
+    dt = tstop-tstart
+    with torch.no_grad():
+      k = torch.norm(t_y).item()
+      q = dt*layer(t_y)
+      kq = torch.norm(q).item()
+      t_y.add_(q)
+      del q
 
-      self.setVectorWeights(tstop,0.0,level,y)
+    #if y.getSendFlag():
```
Comment (steffi7574) [inline torchbraid/odenet_apps.py:174]: Comment can be removed.

File: torchbraid/odenet_apps.py [hunk 10]
```diff
@@ -230,20 +193,22 @@ def getPrimalWithGrad(self,tstart,tstop,level):
...
-    
-    layer = self.getLayer(tstart,tstop,level)
 
     b_x = self.getUVector(0,tstart)
-    t_x = b_x.tensor()
 
-    self.setLayerWeights(tstart,tstop,level,b_x.weightTensors())
+    ts_index = self.getGlobalTimeIndex(tstart)-self.start_layer
 
+    assert(ts_index<len(self.layer_models))
```
Comment (steffi7574) [inline torchbraid/odenet_apps.py:201]: Might be good to have an assert(ts_index >= 0) here too, just to be sure, and catch potential future bugs...

**Review by steffi7574 — COMMENTED** [review]
Great! I think this is good now, using the correct layers in my tests. I commented on a few parts where I believe that removing obsolete comments would be helpful. Otherwise this is good to be merged I think. (I couldn't review the rnn_apps.py though).

**Comment by eric-c-cyr** [comment]
Thanks, all of this is done!

From: Github Notifications ***@***.***>
Reply-To: Multilevel-NN/torchbraid ***@***.***>
Date: Tuesday, June 15, 2021 at 11:42 AM
To: Multilevel-NN/torchbraid ***@***.***>
Cc: "Cyr, Eric C" ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [Multilevel-NN/torchbraid] Refactor odenets (#3)


@steffi7574 commented on this pull request.

Great! I think this is good now, using the correct layers in my tests. I commented on a few parts where I believe that removing obsolete comments would be helpful. Otherwise this is good to be merged I think. (I couldn't review the rnn_apps.py though).

________________________________

In torchbraid/odenet_apps.py<https://github.com/Multilevel-NN/torchbraid/pull/3#discussion_r652010954>:

> -      # get some information about what to do

-      dt = tstop-tstart

-      layer = self.getLayer(tstart,tstop,level) # resnet "basic block"

-

-      #print(self.my_rank, ": FWDeval level ", level, " ", tstart, "->", tstop, " using layer ", layer.getID(), ": ", layer.linearlayer.weight[0].data)

-

-      if t_x==None:

-        t_x = t_y

-      else:

-        t_y.copy_(t_x)

-

-      q = dt*layer(t_x)

-      t_y.add_(q)

-      del q

-    # end in_place_eval

-

     # there are two paths by which eval is called:

This comments is confusing now, and probably should be removed. As far as I see it, there is not anymore two path which call eval, right?

________________________________

In torchbraid/odenet_apps.py<https://github.com/Multilevel-NN/torchbraid/pull/3#discussion_r652011872>:

>

-      self.setVectorWeights(tstop,0.0,level,y)

+    #if y.getSendFlag():

Comment can be removed.

________________________________

In torchbraid/odenet_apps.py<https://github.com/Multilevel-NN/torchbraid/pull/3#discussion_r652012300>:

>

+    assert(ts_index<len(self.layer_models))

Might be good to have an assert(ts_index >= 0) here too, just to be sure, and catch potential future bugs...

**Pull request merged**